ConstraintSystem: replace data structure with Matrix

artagnon opened the pull request:
ConstraintSystem currently stores constraints in a vector-of-vectors, performing inefficient memory operations on it, as part of its operation. Replace this data structure with the newly-minted Matrix, using JaggedArrayView to eliminate row-swaps and truncation of the underlying storage.

Since Matrix requires knowing an upper bounds on the number of columns ahead-of-time, use the newly-added dry-run routine  to do the entire allocation upfront.

-- 8< --
Based on #98893 and #99670.

artagnon committed:
ADT/Matrix: two-dimensional Container with View

Leverage the excellent SmallVector infrastructure to write a
two-dimensional container, along with a View that abstracts out
indexing-arithmetic, eliminating memory operations on the underlying
storage. The immediate applicability of Matrix is to replace uses of the
vector-of-vectors idiom, with one caveat: an upper bound on the number
of columns should be known ahead of time.

artagnon committed:
ConstraintSystem: replace data structure with Matrix

ConstraintSystem currently stores constraints in a vector-of-vectors,
performing inefficient memory operations on it, as part of its
operation. Replace this data structure with the newly-minted Matrix,
using MatrixView to eliminate row-swaps and truncation of the underlying
storage.

Since Matrix requires knowing an upper bounds on the number of columns
ahead-of-time, add a cl::opt to ConstraintElimination upper-bounding
this, and change addVariableRowFill to not add constraints whose length
exceeds this upper bound.

arsenm commented on llvm/unittests/ADT/MatrixTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  auto &M = this->SmallMatrix;
+  auto &O = this->OtherSmall;
+  MatrixView<TypeParam> V{M};
+  ASSERT_EQ(M.getNumRows(), 2u);
```
Most of these ASSERT_EQ should be EXPECT_EQ

artagnon replied:
Fixed in #98893. Will rebase after dependent patches are fully reviewed.

artagnon commented:
The LLVM Compile Time Tracker results are in, and they look good: https://llvm-compile-time-tracker.com/compare.php?from=17844494b4ea0bc465fc31b745a537b9367d6608&to=b1c870b25bb7fa20a17362f3e14557830bb31ff2&stat=instructions:u.

fhahn commented:
Thanks for working on improving the representation used!

> Since Matrix requires knowing an upper bounds on the number of columns ahead-of-time, add a cl::opt to ConstraintElimination upper-bounding this, and change addVariableRowFill to not add constraints whose length exceeds this upper bound.

I think the current cutoff is probably too low, I think in practice we often have systems with more than 16 columns/variables (there are a number of binary changes in CTMark due to this I think https://llvm-compile-time-tracker.com/compare.php?from=ef1851a4daa5a9b48cdff9bac12a116c6518d39b&to=f78bc6008c7dea8acce3a4171dd3ece15103419f&stat=size-text).

To compare apples to apples compile-time wise, it would be good to pick the default so we handle all cases handled without this change (might pessimize this change due to large up-front allocations though) or possibly resize the matrix on demand (which would need to repack the existing data)

artagnon commented:
> I think the current cutoff is probably too low, I think in practice we often have systems with more than 16 columns/variables (there are a number of binary changes in CTMark due to this I think https://llvm-compile-time-tracker.com/compare.php?from=ef1851a4daa5a9b48cdff9bac12a116c6518d39b&to=f78bc6008c7dea8acce3a4171dd3ece15103419f&stat=size-text).

I see. Should I attempt to change it to 32 or 64? What's a good upper-bound on the number of columns, if the number of rows are upper-bounded at 500?

> To compare apples to apples compile-time wise, it would be good to pick the default so we handle all cases handled without this change (might pessimize this change due to large up-front allocations though) or possibly resize the matrix on demand (which would need to repack the existing data)

I think one large upfront allocation wouldn't hurt as much as repeatedly calling malloc/free to resize the underlying storage. I think things can be improved by doing a dry-run, upper-bounding the number of rows and columns, and performing the entire allocation upfront.

artagnon commented:
> I think one large upfront allocation wouldn't hurt as much as repeatedly calling malloc/free to resize the underlying storage. I think things can be improved by doing a dry-run, upper-bounding the number of rows and columns, and performing the entire allocation upfront.

Another thing I noticed was a call to a copy-constructor that copies the underlying storage. Is it really necessary to copy the underlying storage?

artagnon committed:
ConstraintElim: add dry-run routine to fail early

Add a dry-run routine that computes a conservative estimate of the
number of rows and columns that the transform will require, and fail
early if the estimates exceed the upper bounds. This patch has a small
overhead, but improves compile-time on one benchmark significantly. The
overhead will be compensated for in a follow-up patch, where
ConstraintSystem is ported to use a Matrix data structure, performing
the full allocation ahead-of-time using these estimates.

fhahn commented:
Looks like Clang crashes with the dry-run version on CTmark https://llvm-compile-time-tracker.com/show_error.php?commit=94bdfca2db58d6e6ff507d073e407acba1046701

artagnon commented:
> Looks like Clang crashes with the dry-run version on CTmark https://llvm-compile-time-tracker.com/show_error.php?commit=94bdfca2db58d6e6ff507d073e407acba1046701

Should be fixed now. However, LLVM Compile Time Tracker times out now, and I have no idea how to go about debugging that: https://llvm-compile-time-tracker.com/show_error.php?commit=82ddf545ab4688a550881a7a7404d40fc585143e.

fhahn commented:
> > Looks like Clang crashes with the dry-run version on CTmark https://llvm-compile-time-tracker.com/show_error.php?commit=94bdfca2db58d6e6ff507d073e407acba1046701
> 
> Should be fixed now. However, LLVM Compile Time Tracker times out now, and I have no idea how to go about debugging that: https://llvm-compile-time-tracker.com/show_error.php?commit=82ddf545ab4688a550881a7a7404d40fc585143e.

This looks like a hang while building clang in stage 2 (while building LLVM/Clang with a clang that contains 82ddf545ab4688a550881a7a7404d40fc585143e ). Doing a stage2 build should hopefully reproduce the issue locally

kuhar commented on llvm/include/llvm/ADT/Matrix.h:
```diff
@@ -0,0 +1,339 @@
+//===- Matrix.h - Two-dimensional Container with View -----------*- C++ -*-===//
```
Would it be possible to first have this code live somewhere local to `ConstraintSystem` and once it matures / gains more usage, promote it to ADT?

artagnon replied:
Possible, since it's header-only. I would like some other opinions as well.

kuhar replied:
I'd strongly prefer this option then. Also, please do not click 'resolve' on conversations that haven't been resolved.

kuhar commented on llvm/include/llvm/ADT/ArrayRef.h:
```diff
@@ -59,6 +60,8 @@ namespace llvm {
     /// The number of elements.
     size_type Length = 0;
 
+    friend MutableRowView<T>;
```
Why do we need this?

artagnon replied:
To access private members Length and Data.

antoniofrighetto commented on llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:
```diff
@@ -1725,6 +1786,74 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
+  SmallVector<Value *> FunctionArgs;
+  for (Value &Arg : F.args())
+    FunctionArgs.push_back(&Arg);
+  auto [S, EstimatedColumns] = dryRun(F, DT, LI, SE);
```
```suggestion
  const auto &[S, EstimatedColumns] = dryRun(F, DT, LI, SE);
```
Then `const FactOrCheck &CB` in the loop below, and `CB` when passed by in `checkOrAndOpImpliedByOther`. `dryRun` is supposed to be the only to modify them.

artagnon replied:
I've used a `const_cast` on CB when passed to `checkOrAndOpImpliedByOther`. Let me know if this is what you meant.

antoniofrighetto replied:
No need to const_cast CB, you can just change `checkOrAndOpImpliedByOther` signature and have a `const FactOrCheck &CB` as first parameter.

fhahn commented:
@artagnon let me know once there's a stable point for the patches, then I can run them on a larger set of programs to check the wider impact

artagnon commented:
> @artagnon let me know once there's a stable point for the patches, then I can run them on a larger set of programs to check the wider impact

Will do, thanks! Sorry about the heavy load of notifications from pushes, while this is still in development: I didn't know to check `size-text` in advance.

artagnon commented:
Okay, the final results from LLVM Compile Time Tracker are back, and it looks like this change has a significant impact on one benchmark, and negligible regressions in ReleaseLTO: http://llvm-compile-time-tracker.com/compare.php?from=c5ee3c05ca61f3fae11337c5853aee7b450a9dc6&to=15dd68506d479cc50349de1cf1e70bd896b9c82e&stat=instructions%3Au.

@fhahn Could you check it on some more programs, when you have the time?
